junit-platform-engine/src/main/java/org/junit/platform/engine/support/discovery/EngineDiscoveryRequestResolution.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright 2015-2019 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.engine.support.discovery;
+
+import static java.util.stream.Collectors.joining;
+import static org.junit.platform.commons.util.BlacklistedExceptions.rethrowIfBlacklisted;
+import static org.junit.platform.commons.util.CollectionUtils.getOnlyElement;
+
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.junit.platform.commons.JUnitException;
+import org.junit.platform.commons.logging.Logger;
+import org.junit.platform.engine.DiscoverySelector;
+import org.junit.platform.engine.EngineDiscoveryRequest;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.discovery.ClassSelector;
+import org.junit.platform.engine.discovery.ClasspathResourceSelector;
+import org.junit.platform.engine.discovery.ClasspathRootSelector;
+import org.junit.platform.engine.discovery.DirectorySelector;
+import org.junit.platform.engine.discovery.FileSelector;
+import org.junit.platform.engine.discovery.MethodSelector;
+import org.junit.platform.engine.discovery.ModuleSelector;
+import org.junit.platform.engine.discovery.PackageSelector;
+import org.junit.platform.engine.discovery.UniqueIdSelector;
+import org.junit.platform.engine.discovery.UriSelector;
+import org.junit.platform.engine.support.discovery.SelectorResolver.Context;
+import org.junit.platform.engine.support.discovery.SelectorResolver.Match;
+import org.junit.platform.engine.support.discovery.SelectorResolver.Resolution;
+
+/**
+ * @since 1.5
+ */
+class EngineDiscoveryRequestResolution {
+
+	private final Logger logger;
+	private final EngineDiscoveryRequest request;
+	private final Context defaultContext;
+	private final List<SelectorResolver> resolvers;
+	private final List<TestDescriptor.Visitor> visitors;
+	private final TestDescriptor engineDescriptor;
+	private final Map<DiscoverySelector, Resolution> resolvedSelectors = new LinkedHashMap<>();
+	private final Map<UniqueId, Match> resolvedUniqueIds = new LinkedHashMap<>();
+	private final Queue<DiscoverySelector> remainingSelectors = new LinkedList<>();
```

Comment:
Have you considered using an `ArrayDeque` instead of a `LinkedList` here?

Reply:
I haven't TBH. Do you think it would make a difference in this scenario performance-wise?

Reply:
I'm not sure, actually. I personally prefer `ArrayDeque` because its Javadoc says that it's "faster than `LinkedList` when used as a queue", so when in doubt, that's the implementation I go for.

But having said this, I don't know how `ArrayDeque` using an array for storage would impact its memory consumption on a long-running instance of the JUnit Platform (as in, I don't know (1) if the queue will need to live for a "long" time, and (2) whether it's possible for the queue to get very long and then empty, causing a memory leak if the array isn't resized back to a small size).

So I'd say that using a `LinkedList` here is perfectly acceptable until we have evidence that it's a performance bottleneck or a memory hog. :)

Reply:
I've now opted for `ArrayDeque` for the reason you mentioned. The queue is short-lived and will only live for the duration of test discovery so we shouldn't have to worry about it being a memory leak.

Reply:
Sounds good to me @marcphilipp! Cheers. 👍